### PR TITLE
Use flake8 for code quality checks

### PR DESCRIPTION
**Pull request by victorlin**
### Description of proposed changes

Flake8 is already used in [nextstrain/cli](https://github.com/nextstrain/cli). It'd be good to use it here too. Some of the commits in this PR should speak to its usefulness (e.g. catching skipped tests due to duplicate names, unused imports).

### Related issue(s)

Prompted by https://github.com/nextstrain/augur/pull/1288#issuecomment-1687161856

Requires #1296 to be merged first

Closes #1288

### Checklist

- [x] Flake8 added to CI (as a pytest)
- [x] Tests updated
- [x] Fix CI issues with Python 3.7 ([thread](https://github.com/nextstrain/augur/pull/1289#discussion_r1309472245))
- [x] Checks pass

### CODE REVIEW

File: .flake8 [hunk 1]
```diff
@@ -0,0 +1,21 @@
...
+extend-ignore =
+    # allow f-strings without any placeholders
+    F541,
+
+    # allow unused variables
+    F841,
+
+per-file-ignores =
+    # allow unused imports for re-export
+    augur/io/__init__.py:F401
```
Comment (victorlin) [inline .flake8:21]: This works locally but [doesn't in CI](https://github.com/nextstrain/augur/actions/runs/5954918209/job/16152462942#step:8:332).

I'll probably resort to inline exceptions.

File: setup.py [hunk 2]
```diff
@@ -70,6 +70,7 @@
         'dev': [
             "cram >=0.7",
             "deepdiff >=4.3.2",
+            "flake8 >=5.0.0",  # 5.0.0 is the latest version that supports Python 3.7
```
Comment (victorlin) [inline setup.py:73]: This change is causing CI's Python 3.7 jobs to [fail](https://github.com/nextstrain/augur/actions/runs/5955308423/job/16327193476?pr=1289) resolving a good Snakemake version.

I'm able to reproduce this locally. Still not sure why the new `flake8` dependency is causing this. `pip install --verbose` doesn't give any more info either.

<details>

<summary>Snippet from output of `pip install -v -e ~/repos/nextstrain/augur'[dev]'`
</summary>

```
Collecting snakemake>=5.4.0 (from nextstrain-augur==22.3.0)
  Obtaining dependency information for snakemake>=5.4.0 from https://files.pythonhosted.org/packages/55/94/a722fc8364fab7c841f6aa1b6eee183ef3d813e381e9d70fb033b1a650b1/snakemake-7.32.2-py3-none-any.whl.metadata
  Using cached snakemake-7.32.2-py3-none-any.whl.metadata (2.7 kB)
  Obtaining dependency information for snakemake>=5.4.0 from https://files.pythonhosted.org/packages/28/34/7088127f6224805d79983bc60a4aadf98402acbdea64186312a9d14a06c1/snakemake-7.32.1-py3-none-any.whl.metadata
  Using cached snakemake-7.32.1-py3-none-any.whl.metadata (2.7 kB)
  Obtaining dependency information for snakemake>=5.4.0 from https://files.pythonhosted.org/packages/2e/03/f925d038363768fe93208de5726ed195ee013939cf471c22b3ca0215c5a1/snakemake-7.32.0-py3-none-any.whl.metadata
  Using cached snakemake-7.32.0-py3-none-any.whl.metadata (2.7 kB)
  Obtaining dependency information for snakemake>=5.4.0 from https://files.pythonhosted.org/packages/cb/84/2065f4618a950dbf7aca49351587bd8f80f7bb58d4bfa594c0b2d2b00a31/snakemake-7.31.1-py3-none-any.whl.metadata
  Using cached snakemake-7.31.1-py3-none-any.whl.metadata (2.7 kB)
  Obtaining dependency information for snakemake>=5.4.0 from https://files.pythonhosted.org/packages/9e/5e/1b6ccd64a7bb6205beb3e5fdd996eabdf6b6cb153007f06203ba2ac78c91/snakemake-7.30.1-py3-none-any.whl.metadata
  Using cached snakemake-7.30.1-py3-none-any.whl.metadata (2.8 kB)
  Obtaining dependency information for snakemake>=5.4.0 from https://files.pythonhosted.org/packages/00/5b/9d492f75e22a533b8bfa53bd3b6007a9d3bae8c6574e46fd01e1efc02911/snakemake-7.30.0-py3-none-any.whl.metadata
  Using cached snakemake-7.30.0-py3-none-any.whl.metadata (2.8 kB)
  Obtaining dependency information for snakemake>=5.4.0 from https://files.pythonhosted.org/packages/ec/d0/645e91d63996d282ed9aaec237b7343934f6e1e2a5399ce1683af06d4d79/snakemake-7.29.0-py3-none-any.whl.metadata
  Using cached snakemake-7.29.0-py3-none-any.whl.metadata (2.8 kB)
  Obtaining dependency information for snakemake>=5.4.0 from https://files.pythonhosted.org/packages/c3/e5/109679e1ec26eb116e3acdcac9d16c8b75d1eb61f581dbd4d873abd71366/snakemake-7.28.3-py3-none-any.whl.metadata
  Using cached snakemake-7.28.3-py3-none-any.whl.metadata (2.8 kB)
  Obtaining dependency information for snakemake>=5.4.0 from https://files.pythonhosted.org/packages/c4/16/25ba3158889d6522c8f509691b6b38b6ffafaaf59df9c6d9ffe13848186d/snakemake-7.28.2-py3-none-any.whl.metadata
  Using cached snakemake-7.28.2-py3-none-any.whl.metadata (2.8 kB)
  Obtaining dependency information for snakemake>=5.4.0 from https://files.pythonhosted.org/packages/9d/f3/ee3fb73548f6b560fa0788192b15bf9c406487ce9f30d8a19b813566f117/snakemake-7.28.1-py3-none-any.whl.metadata
  Using cached snakemake-7.28.1-py3-none-any.whl.metadata (2.8 kB)
  Obtaining dependency information for snakemake>=5.4.0 from https://files.pythonhosted.org/packages/6b/57/ef6f55d9f04ea5f732bcbfc830e4629e4dcc29cc54300b85b7f6d6a64a1d/snakemake-7.28.0-py3-none-any.whl.metadata
  Using cached snakemake-7.28.0-py3-none-any.whl.metadata (2.8 kB)
Collecting msrest (from snakemake>=5.4.0->nextstrain-augur==22.3.0)
  Using cached msrest-0.7.1-py3-none-any.whl (85 kB)
Collecting snakemake>=5.4.0 (from nextstrain-augur==22.3.0)
  Obtaining dependency information for snakemake>=5.4.0 from https://files.pythonhosted.org/packages/d7/66/4a60598821775f108a318064425a250420931ec2ab4619de2a36b274e5e2/snakemake-7.26.0-py3-none-any.whl.metadata
  Using cached snakemake-7.26.0-py3-none-any.whl.metadata (2.5 kB)
  Obtaining dependency information for snakemake>=5.4.0 from https://files.pythonhosted.org/packages/82/f4/00b680935e475d1475e68e9999594bdbcd71b013caaf425455fa55e123b4/snakemake-7.25.4-py3-none-any.whl.metadata
  Using cached snakemake-7.25.4-py3-none-any.whl.metadata (2.5 kB)
  Using cached snakemake-7.25.3-py3-none-any.whl (366 kB)
…
```

Comment (victorlin) [reply setup.py]: Still haven't gotten this working. I've tried:

1. [Pinning Flake8 to an earlier version](https://github.com/nextstrain/augur/pull/1289/commits/50d343055dd9de54ba3bbf5936e1ba4cb0a90e9c).
2. [Unpinning Flake8](https://github.com/nextstrain/augur/pull/1289/commits/ce8c7ae17b298867be5bbfb2de8c0dc8bd04b28c).
3. [Removing Snakemake as a dependency](https://github.com/nextstrain/augur/issues/753) (I did this locally and both `pytest_mock` and `pytest_cov` weren't resolvable either).

I'll note that this is only an issue on Python 3.7, which has reached EOL. The easy solution would be to just drop Python 3.7 support.

Comment (victorlin) [reply setup.py]: [Updated](https://github.com/nextstrain/augur/pull/1289#event-10236208134) to base this PR on #1296.

**Review by huddlej — APPROVED** [review]
This is excellent and something I've wanted to see for a while. Thank you, @victorlin!

I like the way flake8 runs as a unit test instead of being a separate GitHub Check on the code, so we can find flake8 errors in our local dev environment, too.

Even though the dependencies issue that arose is separate from the goal of this PR, we should really plan to remove Snakemake as a dev dependency in a separate PR. My original hope was that cram tests would replace the original Snakemake build tests, but the newer pathogen-repo-ci checks seem to fill the original goal of those Snakemake tests instead. We could probably add a pathogen-repo-ci for TB, port [the remaining builds tests](https://github.com/nextstrain/augur/tree/master/tests/builds) to cram, and then drop the Snakemake dep.

**Comment by corneliusroemer** [comment]
Excellent, thanks @victorlin for setting up flake8 so that things stay clean in the future!

**Pull request merged**